core/src/test/java/org/elasticsearch/script/expression/ExpressionScriptTests.java

```diff
@@ -385,8 +400,150 @@ public void testStringSpecialValueVariable() throws Exception {
-                   message.contains("ExpressionScriptExecutionException"), equalTo(true));
+                message.contains("ExpressionScriptExecutionException"), equalTo(true));
         assertThat(message + "should have contained text variable error",
-                   message.contains("text variable"), equalTo(true));
+                message.contains("text variable"), equalTo(true));
+    }
+
+    public void testExecutableScripts() throws Exception {
+        // series of unit test for using expressions as executable scripts
+        Map<String, Object> vars = new HashMap<>();
+        vars.put("a", 2.5);
+        vars.put("b", 3);
+        vars.put("xyz", -1);
+
+        Expression expr = JavascriptCompiler.compile("a+b+xyz");
+
+        ExpressionExecutableScript ees = new ExpressionExecutableScript(expr, vars);
+        assertEquals((Double) ees.run(), 4.5, 0.001);
+
+        ees.setNextVar("b", -2.5);
+        assertEquals((Double) ees.run(), -1, 0.001);
+
+        ees.setNextVar("a", -2.5);
+        ees.setNextVar("b", -2.5);
+        ees.setNextVar("xyz", -2.5);
+        assertEquals((Double) ees.run(), -7.5, 0.001);
+
+        String message;
+
+        try {
+            vars = new HashMap<>();
+            vars.put("a", 1);
+            ees = new ExpressionExecutableScript(expr, vars);
+            ees.run();
+            fail("An incorrect number of variables were allowed to be used in an expression.");
+        } catch (ScriptException se) {
+            message = se.getMessage();
+            assertThat(message + " should have contained number of variables", message.contains("number of variables"), equalTo(true));
+        }
+
+        try {
+            vars = new HashMap<>();
+            vars.put("a", 1);
+            vars.put("b", 3);
+            vars.put("c", -1);
+            ees = new ExpressionExecutableScript(expr, vars);
+            ees.run();
+            fail("A variable was allowed to be set that does not exist in the expression.");
+        } catch (ScriptException se) {
+            message = se.getMessage();
+            assertThat(message + " should have contained does not exist in", message.contains("does not exist in"), equalTo(true));
+        }
+
+        try {
+            vars = new HashMap<>();
+            vars.put("a", 1);
+            vars.put("b", 3);
+            vars.put("xyz", "hello");
+            ees = new ExpressionExecutableScript(expr, vars);
+            ees.run();
+            fail("A non-number was allowed to be use in the expression.");
+        } catch (ScriptException se) {
+            message = se.getMessage();
+            assertThat(message + " should have contained process numbers", message.contains("process numbers"), equalTo(true));
+        }
+
+        createIndex("test_index", "agg_index");
+        ensureGreen("test_index", "agg_index");
+
+        // test to make sure expressions are not allowed to be used as update scripts
```

Comment:
can we put each of these with comments into different test methods? At least separate the unit test style ones above from these real integration tests?